### PR TITLE
Disable editing the provisioned user's attribute values

### PR DESCRIPTION
**Pull request by GANGANI**
### Description
With the PR changes, editing the provisioned users synced attribute will be disabled based on the config introduced at [1]

Related issue: https://github.com/wso2/product-is/issues/12010

### Action items
This PR should be merged after merging [1] and upgrading the framework version.

[1] https://github.com/wso2/carbon-identity-framework/pull/3589

### CODE REVIEW

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/utils/SCIMCommonConstants.java [hunk 1]
```diff
@@ -83,6 +83,10 @@
 
     public static final String READ_ONLY_USER_CLAIM = "http://wso2.org/claims/identity/isReadOnlyUser";
 
+    public static final String PROVISIONED_SOURCE_CLAIM = "http://wso2.org/claims/identity/userSource";
+    public static final String SYNCED_ATTRIBUTE_EDITING_ENABLED = "JITProvisioning.EnableSyncedAttributeEditing";
+    public static final String PROVISIONED_USER = "provisionedUser";
```
Comment (somindatommy) [inline components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/utils/SCIMCommonConstants.java:88]: Is this the same as the contant in framework? if so why can't we use the same claim?

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/utils/SCIMCommonConstants.java [hunk 2]
```diff
@@ -83,6 +83,10 @@
 
     public static final String READ_ONLY_USER_CLAIM = "http://wso2.org/claims/identity/isReadOnlyUser";
 
+    public static final String PROVISIONED_SOURCE_CLAIM = "http://wso2.org/claims/identity/userSource";
+    public static final String SYNCED_ATTRIBUTE_EDITING_ENABLED = "JITProvisioning.EnableSyncedAttributeEditing";
```
Comment (somindatommy) [inline components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/utils/SCIMCommonConstants.java:87]: Is there any single place we keep JIT provisioning-related configs?

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/listener/SCIMUserOperationListener.java [hunk 3]
```diff
@@ -231,6 +238,56 @@ public boolean doPreSetUserClaimValuesWithID(String userID, Map<String, String>
...
+         */
+        for (Claim claim : userClaimValues) {
+            String claimURI = claim.getClaimUri();
+            if (!PROVISIONED_SOURCE_CLAIM.equals(claimURI)) {
+                continue;
+            }
+            if (StringUtils.isNotBlank(claim.getValue())) {
+                String errorMsg = "The user: " + username + " has been provisioned from federated IDP: " +
+                        claim.getValue() + ". Hence provisioned user attributes are not allowed to update";
+                throw new UserStoreClientException(errorMsg);
```
Comment (somindatommy) [inline components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/listener/SCIMUserOperationListener.java:284]: Any chance we can throw an error code?

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/utils/SCIMCommonConstants.java [hunk 4]
```diff
@@ -83,6 +83,10 @@
 
     public static final String READ_ONLY_USER_CLAIM = "http://wso2.org/claims/identity/isReadOnlyUser";
 
+    public static final String PROVISIONED_SOURCE_CLAIM = "http://wso2.org/claims/identity/userSource";
```
Comment (mevan-karu) [inline components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/utils/SCIMCommonConstants.java:86]: `http://wso2.org/claims/identity/userSource` claim couldn't be used for this improvement. This claim is already used to keep the user store of the user in Asgardeo.

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/listener/SCIMUserOperationListener.java [hunk 5]
```diff
@@ -231,6 +238,56 @@ public boolean doPreSetUserClaimValuesWithID(String userID, Map<String, String>
...
+            } else {
+                break;
+            }
```
Comment (mevan-karu) [inline components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/listener/SCIMUserOperationListener.java:287]:
```suggestion
            } 
            break;
      
```

**Comment by ashensw** [comment]
@GANGANI Can you check the PR Builder failure as well?

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/listener/SCIMUserOperationListener.java [hunk 6]
```diff
@@ -231,6 +238,56 @@ public boolean doPreSetUserClaimValuesWithID(String userID, Map<String, String>
...
+        }
+
+        String username = ((UniqueIDJDBCUserStoreManager) userStoreManager).getUserNameFromUserID(userID);
+        Claim[] userClaimValues = userStoreManager.getUserClaimValues(username, profileName);
+
+        /*
+        If the PROVISIONED_SOURCE_CLAIM is already set as claim of the user, that user can be identified as provisioned
+        user. Therefore attribute editing will be blocked for that user.
+         */
+        for (Claim claim : userClaimValues) {
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/listener/SCIMUserOperationListener.java:276]: let's add a null check for userClaimValues

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/listener/SCIMUserOperationListener.java (hunk 3 above)
Comment (GANGANI) [reply components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/listener/SCIMUserOperationListener.java]: Fixed https://github.com/wso2-extensions/identity-inbound-provisioning-scim2/pull/367/commits/9d70788d485e475ac471678bec53d31f24aca382

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/listener/SCIMUserOperationListener.java (hunk 5 above)
Comment (GANGANI) [reply components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/listener/SCIMUserOperationListener.java]: Fixed https://github.com/wso2-extensions/identity-inbound-provisioning-scim2/pull/367/commits/9d70788d485e475ac471678bec53d31f24aca382

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/listener/SCIMUserOperationListener.java (hunk 6 above)
Comment (GANGANI) [reply components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/listener/SCIMUserOperationListener.java:280]: Fixed https://github.com/wso2-extensions/identity-inbound-provisioning-scim2/pull/367/commits/9d70788d485e475ac471678bec53d31f24aca382

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/utils/SCIMCommonConstants.java (hunk 4 above)
Comment (GANGANI) [reply components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/utils/SCIMCommonConstants.java:86]: A new claim is expected to introduce and will update the PR with that claim URI before merging this.

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/utils/SCIMCommonConstants.java (hunk 2 above)
Comment (GANGANI) [reply components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/utils/SCIMCommonConstants.java]: JIT provisioning related configs are maintained here[1].

[1] https://github.com/wso2/carbon-identity-framework/blob/master/features/identity-core/org.wso2.carbon.identity.core.server.feature/resources/identity.xml.j2#L1266-L1269

**Comment by GANGANI** [comment]
> @GANGANI Can you check the PR Builder failure as well?

This PR depends on [1] as a constant is referred from there. I will merge [1] and bump the version before merging this.

[1] https://github.com/wso2/carbon-identity-framework/pull/3589

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/listener/SCIMUserOperationListener.java [hunk 7]
```diff
@@ -221,6 +227,8 @@ public boolean doPreSetUserClaimValuesWithID(String userID, Map<String, String>
             throw new UserStoreException("Error while reading isScimEnabled from userstore manager", e);
         }
 
+        validateProvisionedUserClaimUpdate(userID, profileName, userStoreManager);
```
Comment (IsuraD) [inline components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/listener/SCIMUserOperationListener.java:230]: Need to samething for doPreSetUserClaimValues, doPreSetUserClaimValue, doPreSetUserClaimValueWithID

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/listener/SCIMUserOperationListener.java [hunk 8]
```diff
@@ -231,6 +239,58 @@ public boolean doPreSetUserClaimValuesWithID(String userID, Map<String, String>
...
+     * @throws UserStoreException if an error occurred while retrieving the user claim list.
+     */
+    private void validateProvisionedUserClaimUpdate(String userID, String profileName,
+                                                    UserStoreManager userStoreManager) throws UserStoreException {
+
+        /*
+        Check whether the config is enabled to edit synced attributes of the provisioned user.
+        If it is enabled, blocking the attribute editing is not required.
+         */
+        if (Boolean.parseBoolean(IdentityUtil.getProperty(SYNCED_ATTRIBUTE_EDITING_ENABLED))) {
```
Comment (IsuraD) [inline components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/listener/SCIMUserOperationListener.java:257]: Use a property name such as, EnableEnhansedJITFeatures

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/listener/SCIMUserOperationListener.java [hunk 9]
```diff
@@ -231,6 +239,58 @@ public boolean doPreSetUserClaimValuesWithID(String userID, Map<String, String>
...
+        /*
+        Check whether this is an attribute syncing flow by checking the PROVISIONED_USER thread local property.
+        If it is an attribute syncing flow, blocking the attribute editing is not required.
+         */
+        if (IdentityUtil.threadLocalProperties.get().get(FrameworkConstants.JIT_PROVISIONED_USER) != null &&
+                (Boolean) IdentityUtil.threadLocalProperties.get().get(FrameworkConstants.JIT_PROVISIONED_USER)) {
+            return;
+        }
+
+        String username = ((UniqueIDJDBCUserStoreManager) userStoreManager).getUserNameFromUserID(userID);
```
Comment (IsuraD) [inline components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/listener/SCIMUserOperationListener.java:270]: why do we need username ? use getUserClaimValuesWithID instead

Comment (IsuraD) [reply components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/listener/SCIMUserOperationListener.java:270]: Also don't retrieve all the claims, just pass the PROVISIONED_SOURCE_CLAIM claim and retrieve it.

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/listener/SCIMUserOperationListener.java [hunk 10]
```diff
@@ -231,6 +239,58 @@ public boolean doPreSetUserClaimValuesWithID(String userID, Map<String, String>
...
+        user. Therefore attribute editing will be blocked for that user.
+         */
+        for (Claim claim : userClaimValues) {
+            String claimURI = claim.getClaimUri();
+            if (!PROVISIONED_SOURCE_CLAIM.equals(claimURI)) {
+                continue;
+            }
+            if (StringUtils.isNotBlank(claim.getValue())) {
+                throw new UserStoreClientException(
+                        SCIMCommonConstants.ErrorMessages.INVALID_PASSWORD_RECOVERY_REQUEST.getCode(),
```
Comment (IsuraD) [inline components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/listener/SCIMUserOperationListener.java:287]: INVALID_PASSWORD_RECOVERY_REQUEST ?

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/utils/SCIMCommonConstants.java [hunk 11]
```diff
@@ -112,5 +115,48 @@
...
             "Date of Birth is not in the correct format of YYYY-MM-DD";
     public static final String DATE_OF_BIRTH_REGEX = "^\\d{4}-\\d{2}-\\d{2}$";
+
+    /**
+     * Enum which contains the error codes and corresponding error messages.
+     */
+    public enum ErrorMessages {
+
+        // SUO - SCIM User Operations.
+        INVALID_PASSWORD_RECOVERY_REQUEST("SUO-10000", "User attribute update is not allowed",
```
Comment (IsuraD) [inline components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/utils/SCIMCommonConstants.java:125]: Change the name INVALID_PASSWORD_RECOVERY_REQUEST

Comment (GANGANI) [reply components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/utils/SCIMCommonConstants.java]: Fixed https://github.com/wso2-extensions/identity-inbound-provisioning-scim2/pull/367/commits/7eb9fee089a75e7a984e91ef3ecb3c2ce5ee165a

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/listener/SCIMUserOperationListener.java (hunk 10 above)
Comment (GANGANI) [reply components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/listener/SCIMUserOperationListener.java]: Fixed https://github.com/wso2-extensions/identity-inbound-provisioning-scim2/pull/367/commits/7eb9fee089a75e7a984e91ef3ecb3c2ce5ee165a

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/listener/SCIMUserOperationListener.java (hunk 9 above)
Comment (GANGANI) [reply components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/listener/SCIMUserOperationListener.java]: Only doPostGetUserClaimValue method[1] was override in IdentityStoreEventListener. Hense only getUserClaimValues method can be used to get identity claim values. Fixed to only retrieve  PROVISIONED_SOURCE_CLAIM value only with https://github.com/wso2-extensions/identity-inbound-provisioning-scim2/pull/367/commits/f4e735453ad2ac354f66a24d63100cecac594701

[1] https://github.com/wso2-extensions/identity-governance/blob/16c7904e76a13b0c317bb8390f3544e958a749fe/components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/listener/IdentityStoreEventListener.java#L213

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/listener/SCIMUserOperationListener.java (hunk 8 above)
Comment (GANGANI) [reply components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/listener/SCIMUserOperationListener.java]: Fixed with https://github.com/wso2-extensions/identity-inbound-provisioning-scim2/pull/367/commits/f4e735453ad2ac354f66a24d63100cecac594701

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/listener/SCIMUserOperationListener.java (hunk 7 above)
Comment (GANGANI) [reply components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/listener/SCIMUserOperationListener.java]: Fixed with https://github.com/wso2-extensions/identity-inbound-provisioning-scim2/pull/367/commits/f4e735453ad2ac354f66a24d63100cecac594701

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/listener/SCIMUserOperationListener.java [hunk 12]
```diff
@@ -231,6 +244,74 @@ public boolean doPreSetUserClaimValuesWithID(String userID, Map<String, String>
...
+        /*
+        Check whether this is an attribute syncing flow by checking the PROVISIONED_USER thread local property.
+        If it is an attribute syncing flow, blocking the attribute editing is not required.
+         */
+        if (IdentityUtil.threadLocalProperties.get().get(FrameworkConstants.JIT_PROVISIONING_FLOW) != null &&
+                (Boolean) IdentityUtil.threadLocalProperties.get().get(FrameworkConstants.JIT_PROVISIONING_FLOW)) {
+            return;
+        }
+
+        boolean isExistingJITProvisionedUser;
```
Comment (IsuraD) [inline components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/listener/SCIMUserOperationListener.java:300]: Although we don't show syncEnabled properly in UI level, can we have a property in BE and check this here? ideally this should happen only if we enabled sync from IDP.

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/listener/SCIMUserOperationListener.java [hunk 13]
```diff
@@ -167,6 +169,11 @@ public boolean doPostSetUserClaimValueWithID(String s, UserStoreManager userStor
     public boolean doPreSetUserClaimValueWithID(String userID, String claimURI, String claimValue, String profileName,
                                                 UserStoreManager userStoreManager) throws UserStoreException {
 
+        // Validate whether claim update request is for a provisioned user.
+        if (isJitProvisionEnhancedFeature()) {
```
Comment (IsuraD) [inline components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/listener/SCIMUserOperationListener.java:173]: isn't this check inside "validateJitProvisionedUserClaimUpdate"

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/listener/SCIMUserOperationListener.java [hunk 14]
```diff
@@ -167,6 +169,11 @@ public boolean doPostSetUserClaimValueWithID(String s, UserStoreManager userStor
...
+            validateJitProvisionedUserClaimUpdate(getUsernameFromUserID(userID, userStoreManager),
```
Comment (IsuraD) [inline components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/listener/SCIMUserOperationListener.java:174]: better rename to validateClaimUpdate or similar

Comment (GANGANI) [reply components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/listener/SCIMUserOperationListener.java]: Fixed https://github.com/wso2-extensions/identity-inbound-provisioning-scim2/pull/367/commits/615427cb98805016d37d6068432a2f558703d653

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/listener/SCIMUserOperationListener.java (hunk 13 above)
Comment (GANGANI) [reply components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/listener/SCIMUserOperationListener.java:173]: Fixed https://github.com/wso2-extensions/identity-inbound-provisioning-scim2/pull/367/commits/615427cb98805016d37d6068432a2f558703d653

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/listener/SCIMUserOperationListener.java (hunk 12 above)
Comment (GANGANI) [reply components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/listener/SCIMUserOperationListener.java:298]: Fixed https://github.com/wso2-extensions/identity-inbound-provisioning-scim2/pull/367/commits/615427cb98805016d37d6068432a2f558703d653

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/listener/SCIMUserOperationListener.java [hunk 15]
```diff
@@ -167,6 +169,10 @@ public boolean doPostSetUserClaimValueWithID(String s, UserStoreManager userStor
     public boolean doPreSetUserClaimValueWithID(String userID, String claimURI, String claimValue, String profileName,
                                                 UserStoreManager userStoreManager) throws UserStoreException {
 
+        // Validate whether claim update request is for a provisioned user.
+        if (isJitProvisionEnhancedFeature()) {
```
Comment (ashensw) [inline components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/listener/SCIMUserOperationListener.java:173]: `isJitProvisionEnhancedFeatureEnabled` is more meaningful IMO. WDYT?

Comment (GANGANI) [reply components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/listener/SCIMUserOperationListener.java]: Fixed https://github.com/wso2-extensions/identity-inbound-provisioning-scim2/pull/367/commits/0652c627f1d971f39b773c2319554686a7f5218b

**Review by ashensw — APPROVED** [review]

**Pull request merged**